File: Valid parenthesis/VP.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        mapping = {')': '(', '}': '{', ']': '['}  # Map closing brackets to their corresponding opening brackets

        for char in s:
            if char in mapping:  # If it's a closing bracket
                top_element = stack.pop() if stack else '#'  # Get the top element or a placeholder
                if mapping[char] != top_element:  # Check if it matches the expected opening bracket
                    return False
            else:  # If it's an opening bracket
                stack.append(char)

        return not stack  # Stack should be empty for the string to be valid


# Test the Solution
if __name__ == "__main__":
    solution = Solution()
    test = "()[]{}"
    if solution.isValid(test):
        print("Valid")
    else:
        print("Invalid")

File: Valid parenthesis/VP.cpp
//Write a fn to find the 

#include <iostream>
#include <stack>
#include <string> // Include string for compatibility
using namespace std;

class Solution {
public:
    bool isValid(string s) {
        stack<char> stack;
        for (const char c : s) {
            if (c == '(') {
                stack.push(')');
            } else if (c == '{') {
                stack.push('}');
            } else if (c == '[') {
                stack.push(']');
            } else if (stack.empty() || stack.top() != c) {
                return false;
            } else {
                stack.pop();
            }
        }
        return stack.empty();
    }
};

int main() {
    Solution solution;
    string test = "()[]{}";
    if (solution.isValid(test)) {
        cout << "Valid" << endl;
    } else {
        cout << "Invalid" << endl;
    }
    return 0;
}
